File: Packages/MiniWhisperCore/Sources/MWStreaming/WebSocketEngine.swift
import AVFAudio
import Foundation
import MWAudio
import MWConfig
import MWSupport
import os

public enum WebSocketEngineError: Error, CustomStringConvertible {
    case preconnectCapExceeded(engine: EngineName, seconds: TimeInterval)

    public var description: String {
        switch self {
        case .preconnectCapExceeded(let engine, let seconds):
            "\(engine.rawValue): connection not open after \(Int(seconds))s of buffered audio"
        }
    }
}

/// A latch that many tasks can await and one can raise, cancellation-safe so the
/// racing task group in `finish` can always unwind.
private final class Latch: @unchecked Sendable {
    private let lock = NSLock()
    private var raised = false
    private var waiters: [UInt64: CheckedContinuation<Void, Never>] = [:]
    private var cancelled: Set<UInt64> = []
    private var nextID: UInt64 = 0

    func raise() {
        let pending = lock.withLock { () -> [CheckedContinuation<Void, Never>] in
            raised = true
            defer { waiters.removeAll() }
            return Array(waiters.values)
        }
        for waiter in pending { waiter.resume() }
    }

    func wait() async {
        let id = lock.withLock { () -> UInt64 in
            nextID += 1
            return nextID
        }
        await withTaskCancellationHandler {
            await withCheckedContinuation { (continuation: CheckedContinuation<Void, Never>) in
                let resumeNow = lock.withLock { () -> Bool in
                    if raised || cancelled.remove(id) != nil { return true }
                    waiters[id] = continuation
                    return false
                }
                if resumeNow { continuation.resume() }
            }
        } onCancel: {
            let waiter = lock.withLock { () -> CheckedContinuation<Void, Never>? in
                guard let waiter = waiters.removeValue(forKey: id) else {
                    cancelled.insert(id)
                    return nil
                }
                return waiter
            }
            waiter?.resume()
        }
    }
}

/// The shared cloud streaming skeleton (F20, F21), a port of
/// `../mini-whisper/src/mini_whisper/streaming/websocket_engine.py`: buffer-until-open
/// with a 60 s cap, send and receive loops, the end-of-audio gate before a terminal
/// event counts, the drain window, single-shot failure and the 5 s finish handshake.
public final class WebSocketEngine<Adapter: EngineAdapter>: StreamingEngine, @unchecked Sendable {
    public static var maxPreconnectSeconds: TimeInterval { 60 }

    private enum QueueItem: Sendable {
        case chunk([Float], Double)
        case end
    }

    private enum FeedOutcome {
        case drop
        case enqueue
        case overflow
    }

    private struct State {
        var adapter: Adapter
        var sink: (any TranscriptSink)?
        var connection: (any WebSocketConnection)?
        var tasks: [Task<Void, Never>] = []
        var open = false
        var finishing = false
        var endSent = false
        var failed = false
        var bufferedSeconds = 0.0
        var fedSeconds = 0.0
        var inputTokens = 0
        var outputTokens = 0
        var assembler = TranscriptAssembler()
    }

    public let name: EngineName

    private let state: OSAllocatedUnfairLock<State>
    private let clock: any Clock
    private let connect: @Sendable () async throws -> any WebSocketConnection
    private let queue: AsyncStream<QueueItem>
    private let queueContinuation: AsyncStream<QueueItem>.Continuation
    private let done = Latch()
    private let log: LogCategory

    public init(
        adapter: Adapter,
        clock: any Clock = SystemClock(),
        connect: @escaping @Sendable () async throws -> any WebSocketConnection
    ) {
        name = adapter.name
        state = OSAllocatedUnfairLock(initialState: State(adapter: adapter))
        self.clock = clock
        self.connect = connect
        (queue, queueContinuation) = AsyncStream.makeStream(bufferingPolicy: .unbounded)
        log = Log.stream(adapter.name.rawValue)
    }

    public convenience init(adapter: Adapter, clock: any Clock = SystemClock()) {
        self.init(adapter: adapter, clock: clock) {
            URLSessionWebSocketConnection(url: adapter.url, headers: adapter.headers)
        }
    }

    deinit {
        for task in state.withLock({ $0.tasks }) { task.cancel() }
        queueContinuation.finish()
    }

    // MARK: - StreamingEngine

    public func start(sink: any TranscriptSink) {
        state.withLock { $0.sink = sink }
        let task = Task { await self.run() }
        state.withLock { $0.tasks.append(task) }
    }

    /// Called synchronously on the audio tap thread: copy the samples out and enqueue.
    public func feed(_ buffer: AVAudioPCMBuffer) {
        guard let channel = buffer.floatChannelData?[0] else { return }
        let count = Int(buffer.frameLength)
        let rate = buffer.format.sampleRate
        guard count > 0, rate > 0 else { return }
        let samples = [Float](UnsafeBufferPointer(start: channel, count: count))
        let seconds = Double(count) / rate

        let outcome = state.withLock { state -> FeedOutcome in
            guard !state.failed, !state.finishing else { return .drop }
            state.fedSeconds += seconds
            guard !state.open else { return .enqueue }
            state.bufferedSeconds += seconds
            return state.bufferedSeconds > Self.maxPreconnectSeconds ? .overflow : .enqueue
        }
        switch outcome {
        case .drop:
            return
        case .enqueue:
            queueContinuation.yield(.chunk(samples, rate))
        case .overflow:
            fail(WebSocketEngineError.preconnectCapExceeded(
                engine: name, seconds: Self.maxPreconnectSeconds
            ))
        }
    }

    public func finish(timeout: Duration) async -> StreamResult {
        state.withLock { $0.finishing = true }
        queueContinuation.yield(.end)
        queueContinuation.finish()

        if await !completedBefore(timeout) {
            log.info("finish timeout after \(timeout.seconds)s")
            state.withLock { $0.failed = true }
        }

        let (tasks, connection) = state.withLock { state -> ([Task<Void, Never>], (any WebSocketConnection)?) in
            defer {
                state.tasks = []
                state.connection = nil
            }
            return (state.tasks, state.connection)
        }
        for task in tasks { task.cancel() }
        connection?.close()

        let snapshot = state.withLock { $0 }
        let usage = StreamUsage(
            inputTokens: snapshot.inputTokens,
            outputTokens: snapshot.outputTokens,
            seconds: snapshot.fedSeconds
        )
        guard !snapshot.failed else { return StreamResult(text: "", ok: false, usage: usage) }
        log.info("session finished")
        return StreamResult(text: snapshot.assembler.text, ok: true, usage: usage)
    }

    // MARK: - Session

    private func run() async {
        let connection: any WebSocketConnection
        do {
            connection = try await connect()
        } catch {
            fail(error)
            return
        }
        let openMessages = state.withLock { state -> [WebSocketMessage] in
            state.connection = connection
            state.open = true
            return state.adapter.openMessages()
        }
        log.info("websocket open")

        do {
            for message in openMessages { try await connection.send(message) }
        } catch {
            fail(error)
            connection.close()
            return
        }

        let receiving = Task { await self.receiveLoop(connection) }
        state.withLock { $0.tasks.append(receiving) }
        await sendLoop(connection)
        if state.withLock({ $0.failed }) { receiving.cancel() }
        await receiving.value
        connection.close()
    }

    private func sendLoop(_ connection: any WebSocketConnection) async {
        var converter: PCMConverter?
        var converterRate: Double?

        for await item in queue {
            guard !state.withLock({ $0.failed }) else { return }
            switch item {
            case .end:
                // Flagged before the send, not after: the receive loop runs
                // concurrently, and a server cannot answer a message it has not
                // received yet — so this cannot let an early terminal through.
                let messages = state.withLock { state -> [WebSocketMessage] in
                    state.endSent = true
                    return state.adapter.endMessages()
                }
                do {
                    for message in messages { try await connection.send(message) }
                } catch {
                    fail(error)
                }
                return
            case .chunk(let samples, let rate):
                if converterRate != rate {
                    converter = PCMConverter(from: rate, to: state.withLock { $0.adapter.targetRate })
                    converterRate = rate
                }
                let pcm = converter?.convert(samples) ?? Data()
                guard !pcm.isEmpty else { continue }
                do {
                    try await connection.send(state.withLock { $0.adapter.encodeChunk(pcm) })
                } catch {
                    fail(error)
                    return
                }
            }
        }
    }

    private func receiveLoop(_ connection: any WebSocketConnection) async {
        while !Task.isCancelled {
            let message: WebSocketMessage
            do {
                message = try await connection.receive()
            } catch is CancellationError {
                return
            } catch {
                fail(error)
                return
            }

            let complete: Bool
            do {
                complete = try apply(message)
            } catch {
                fail(error)
                return
            }

            // Terminal only once our end-of-audio message is on the wire; a
            // server-VAD completion processed earlier must not end the loop.
            guard complete, state.withLock({ $0.endSent }) else { continue }
            let window = state.withLock { $0.adapter.drainAfterComplete }
            if window > .zero { await drain(connection, within: window) }
            done.raise()
            return
        }
    }

    /// Scoops events still in flight for up to `window` after the terminal event
    /// (design §5.4).
    private func drain(_ connection: any WebSocketConnection, within window: Duration) async {
        let deadline = clock.now + window.seconds
        while true {
            let remaining = deadline - clock.now
            guard remaining > 0 else { return }
            let received = await withTaskGroup(of: WebSocketMessage?.self) { group in
                group.addTask { try? await connection.receive() }
                group.addTask {
                    try? await self.clock.sleep(for: .seconds(remaining))
                    return nil
                }
                let first = await group.next() ?? nil
                group.cancelAll()
                return first
            }
            guard let received, (try? apply(received)) != nil else { return }
        }
    }

    private func completedBefore(_ timeout: Duration) async -> Bool {
        await withTaskGroup(of: Bool.self) { group in
            group.addTask {
                await self.done.wait()
                return true
            }
            group.addTask {
                try? await self.clock.sleep(for: timeout)
                return false
            }
            let first = await group.next() ?? false
            group.cancelAll()
            return first
        }
    }

    /// Hands one server message to the adapter, then applies its emissions to the
    /// compound transcript and the sink in order.
    private func apply(_ message: WebSocketMessage) throws -> Bool {
        typealias Applied = (complete: Bool, sink: (any TranscriptSink)?, emissions: [AdapterEmitter.Emission])
        let applied = try state.withLock { state -> Applied in
            var emitter = AdapterEmitter()
            let complete = try state.adapter.handle(message, emit: &emitter)
            state.inputTokens += emitter.inputTokens
            state.outputTokens += emitter.outputTokens
            for emission in emitter.emissions {
                switch emission {
                case .partial(let text): state.assembler.addPartial(text)
                case .final(let text): state.assembler.addFinal(text)
                }
            }
            return (complete, state.sink, emitter.emissions)
        }
        for emission in applied.emissions {
            switch emission {
            case .partial(let text): applied.sink?.onPartial(text)
            case .final(let text): applied.sink?.onFinal(text)
            }
        }
        return applied.complete
    }

    /// Single-shot: only the first failure reaches the sink, and the sink hears about
    /// it before `finish` is released.
    private func fail(_ error: any Error) {
        let (report, sink) = state.withLock { state -> (Bool, (any TranscriptSink)?) in
            guard !state.failed else { return (false, nil) }
            state.failed = true
            return (true, state.sink)
        }
        guard report else { return }
        log.info("engine failed: \(AnyError(error).description)")
        sink?.onEngineError(error)
        done.raise()
    }
}

File: src/mini_whisper/streaming/websocket_engine.py
"""Shared cloud streaming-engine skeleton over websockets, plus subclasses.

The skeleton owns the background thread + asyncio loop, the buffer-until-open
queue (capped at MAX_PRECONNECT_SECONDS of audio), single-shot error
propagation, and the finish handshake with timeout. Subclasses supply only:
name, target_rate, _url(), _headers(), _open_messages(), _encode_chunk(),
_end_messages(), and _handle_event() (design §5).
"""

import asyncio
import base64
import json
import logging
import queue
import threading

import websockets

from mini_whisper.recorder import _buffer_to_numpy
from mini_whisper.streaming.audio_convert import convert, make_state
from mini_whisper.streaming.base import CompoundTranscript, StreamResult

logger = logging.getLogger(__name__)

MAX_PRECONNECT_SECONDS = 60.0

OPENAI_REALTIME_URL = "wss://api.openai.com/v1/realtime"
OPENAI_REALTIME_MODEL = "gpt-live-transcribe"
ELEVENLABS_REALTIME_URL = (
    "wss://api.elevenlabs.io/v1/speech-to-text/realtime"
    "?model_id=scribe_v2_realtime&audio_format=pcm_16000"
)
SPEECHMATICS_REALTIME_URL = "wss://eu.rt.speechmatics.com/v2"


class WebSocketEngine:
    """Base class for the cloud streaming engines. Not an engine by itself."""

    name = ""
    target_rate = 16000
    # Quiet window (s) to scoop straggler events after the terminal one; for
    # providers whose terminal event may race trailing transcript events.
    drain_after_complete = 0.0

    def __init__(self, api_key: str, connect=None):
        self._api_key = api_key
        # Injection seam for tests: async callable returning a websocket-like
        # object with async send/recv/close.
        self._connect = connect or self._default_connect
        self._sink = None
        self._compound = CompoundTranscript()
        self._queue: queue.Queue = queue.Queue()  # (samples, rate) | None = end
        self._thread: threading.Thread | None = None
        self._loop: asyncio.AbstractEventLoop | None = None
        self._open = threading.Event()
        self._finishing = threading.Event()
        self._end_sent = threading.Event()  # end-of-audio message on the wire
        self._done = threading.Event()  # terminal event received, or failed
        self._failed = threading.Event()
        self._fail_lock = threading.Lock()
        self._buffered_seconds = 0.0  # audio queued while not yet open
        self._fed_seconds = 0.0
        self._input_tokens = 0
        self._output_tokens = 0

    # -- public lifecycle (StreamingEngine protocol) -------------------------

    def start(self, sink) -> None:
        self._sink = sink
        self._thread = threading.Thread(
            target=self._thread_main, name=f"stream-{self.name}", daemon=True
        )
        self._thread.start()

    def feed(self, pcm_buffer) -> None:
        if self._failed.is_set() or self._finishing.is_set():
            return
        try:
            samples, rate = self._extract(pcm_buffer)
        except Exception:
            logger.exception("%s: buffer extraction failed", self.name)
            return
        self._fed_seconds += len(samples) / rate
        if not self._open.is_set():
            self._buffered_seconds += len(samples) / rate
            if self._buffered_seconds > MAX_PRECONNECT_SECONDS:
                self._fail(RuntimeError(
                    f"{self.name}: connection not open after "
                    f"{MAX_PRECONNECT_SECONDS:.0f}s of buffered audio"
                ))
                return
        self._queue.put_nowait((samples, rate))

    def finish(self, timeout: float = 5.0) -> StreamResult:
        self._finishing.set()
        self._queue.put_nowait(None)
        finished = self._done.wait(timeout)
        if not finished:
            logger.info("%s: finish timeout after %.1fs", self.name, timeout)
            self._failed.set()
        self._stop_loop()
        usage = {
            "input_tokens": self._input_tokens,
            "output_tokens": self._output_tokens,
            "seconds": self._fed_seconds,
        }
        if self._failed.is_set():
            return StreamResult(text="", ok=False, usage=usage)
        logger.info("%s: session finished", self.name)
        return StreamResult(text=self._compound.text, ok=True, usage=usage)

    # -- engine thread --------------------------------------------------------

    def _thread_main(self) -> None:
        try:
            asyncio.run(self._run())
        except asyncio.CancelledError:
            pass
        except Exception as exc:
            self._fail(exc)

    async def _run(self) -> None:
        self._loop = asyncio.get_running_loop()
        try:
            ws = await self._connect()
        except Exception as exc:
            self._fail(exc)
            return
        self._open.set()
        logger.info("%s: websocket open", self.name)
        try:
            for message in self._open_messages():
                await ws.send(message)
            recv_task = asyncio.create_task(self._recv_loop(ws))
            try:
                await self._send_loop(ws)
                await recv_task
            finally:
                if not recv_task.done():
                    recv_task.cancel()
        except Exception as exc:
            self._fail(exc)
        finally:
            try:
                await ws.close()
            except Exception:
                pass

    async def _send_loop(self, ws) -> None:
        loop = asyncio.get_running_loop()
        state = None
        state_rate = None
        try:
            while not self._failed.is_set():
                item = await loop.run_in_executor(None, self._queue.get)
                if item is None:
                    for message in self._end_messages():
                        await ws.send(message)
                    self._end_sent.set()
                    return
                samples, rate = item
                if state is None or rate != state_rate:
                    state = make_state(rate, self.target_rate)
                    state_rate = rate
                pcm, state = convert(samples, state)
                if pcm:
                    await ws.send(self._encode_chunk(pcm))
        except asyncio.CancelledError:
            raise
        except Exception as exc:
            self._fail(exc)

    async def _recv_loop(self, ws) -> None:
        while True:
            try:
                raw = await ws.recv()
                complete = self._handle_event(raw)
            except asyncio.CancelledError:
                raise
            except Exception as exc:
                self._fail(exc)
                return
            # Terminal only once our end-of-audio message is on the wire; a
            # server-VAD "completed" processed earlier must not end the loop.
            if complete and self._end_sent.is_set():
                if self.drain_after_complete:
                    await self._drain(ws, self.drain_after_complete)
                self._done.set()
                return

    async def _drain(self, ws, quiet: float) -> None:
        """Scoop events still in flight until the socket is quiet."""
        while True:
            try:
                raw = await asyncio.wait_for(ws.recv(), quiet)
            except (TimeoutError, asyncio.TimeoutError):
                return
            except asyncio.CancelledError:
                raise
            except Exception:
                return
            try:
                self._handle_event(raw)
            except Exception:
                return

    def _stop_loop(self) -> None:
        loop = self._loop
        if loop is None or not loop.is_running():
            return

        def _cancel_all():
            for task in asyncio.all_tasks():
                task.cancel()

        loop.call_soon_threadsafe(_cancel_all)

    async def _default_connect(self):
        return await websockets.connect(self._url(), additional_headers=self._headers())

    # -- helpers for subclasses ----------------------------------------------

    @staticmethod
    def _extract(pcm_buffer):
        """AVAudioPCMBuffer → (float32 samples copy, hardware sample rate)."""
        return _buffer_to_numpy(pcm_buffer), float(pcm_buffer.format().sampleRate())

    def _emit_partial(self, text: str) -> None:
        self._compound.add_partial(text)
        if self._sink is not None:
            self._sink.on_partial(text)

    def _emit_final(self, text: str) -> None:
        self._compound.add_final(text)
        if self._sink is not None:
            self._sink.on_final(text)

    def _add_tokens(self, input_tokens, output_tokens) -> None:
        self._input_tokens += int(input_tokens or 0)
        self._output_tokens += int(output_tokens or 0)

    def _fail(self, exc: Exception) -> None:
        with self._fail_lock:
            if self._failed.is_set():
                return
            self._failed.set()
        logger.info("%s: engine failed: %s", self.name, exc)
        self._done.set()
        if self._sink is not None:
            try:
                self._sink.on_engine_error(exc)
            except Exception:
                logger.exception("sink.on_engine_error raised")

    # -- protocol hooks (subclass responsibilities) ---------------------------

    def _url(self) -> str:
        raise NotImplementedError

    def _headers(self) -> dict:
        raise NotImplementedError

    def _open_messages(self) -> list:
        raise NotImplementedError

    def _encode_chunk(self, pcm: bytes):
        raise NotImplementedError

    def _end_messages(self) -> list:
        raise NotImplementedError

    def _handle_event(self, raw) -> bool:
        """Parse one server message; return True when the stream is complete."""
        raise NotImplementedError


class OpenAIRealtimeEngine(WebSocketEngine):
    """OpenAI Realtime transcription session (design §4, pinned 2026-09-01)."""

    name = "openai"
    target_rate = 24000
    # Server VAD can complete a mid-stream segment concurrently with our
    # commit; drain briefly so the post-commit segment is never dropped.
    drain_after_complete = 0.2

    def __init__(self, api_key: str, connect=None):
        super().__init__(api_key, connect=connect)
        self._delta_acc = ""  # deltas accumulate per segment

    def _url(self) -> str:
        return OPENAI_REALTIME_URL

    def _headers(self) -> dict:
        return {"Authorization": f"Bearer {self._api_key}"}

    def _open_messages(self) -> list:
        return [json.dumps({
            "type": "session.update",
            "session": {
                "type": "transcription",
                "audio": {
                    "input": {
                        "format": {"type": "audio/pcm", "rate": self.target_rate},
                        "transcription": {"model": OPENAI_REALTIME_MODEL},
                        "turn_detection": {"type": "server_vad"},
                    }
                },
            },
        })]

    def _encode_chunk(self, pcm: bytes) -> str:
        return json.dumps({
            "type": "input_audio_buffer.append",
            "audio": base64.b64encode(pcm).decode("ascii"),
        })

    def _end_messages(self) -> list:
        return [json.dumps({"type": "input_audio_buffer.commit"})]

    def _handle_event(self, raw) -> bool:
        event = json.loads(raw)
        etype = event.get("type", "")
        if etype == "conversation.item.input_audio_transcription.delta":
            self._delta_acc += event.get("delta", "")
            self._emit_partial(self._delta_acc)
        elif etype == "conversation.item.input_audio_transcription.completed":
            self._delta_acc = ""
            self._emit_final(event.get("transcript", ""))
            usage = event.get("usage") or {}
            self._add_tokens(usage.get("input_tokens"), usage.get("output_tokens"))
            return True
        elif etype == "error":
            raise RuntimeError(f"openai realtime error: {event.get('error')}")
        return False


class ElevenLabsEngine(WebSocketEngine):
    """ElevenLabs Scribe v2 Realtime websocket (design §4)."""

    name = "elevenlabs"
    target_rate = 16000
    # A VAD-committed segment can race our committing chunk, same as OpenAI.
    drain_after_complete = 0.2

    def _url(self) -> str:
        return ELEVENLABS_REALTIME_URL

    def _headers(self) -> dict:
        return {"xi-api-key": self._api_key}

    def _open_messages(self) -> list:
        return []  # session config travels in the URL query parameters

    def _encode_chunk(self, pcm: bytes) -> str:
        return json.dumps({
            "message_type": "input_audio_chunk",
            "audio_base_64": base64.b64encode(pcm).decode("ascii"),
            "commit": False,
            "sample_rate": self.target_rate,
        })

    def _end_messages(self) -> list:
        return [json.dumps({
            "message_type": "input_audio_chunk",
            "audio_base_64": "",
            "commit": True,
            "sample_rate": self.target_rate,
        })]

    def _handle_event(self, raw) -> bool:
        event = json.loads(raw)
        mtype = event.get("message_type", "")
        if mtype in ("partial_transcript", "final_transcript"):
            # final_transcript is settled but not yet committed: still partial
            self._emit_partial(event.get("text", ""))
        elif mtype == "committed_transcript":
            self._emit_final(event.get("text", ""))
            return True
        elif mtype == "input_error":
            raise RuntimeError(f"elevenlabs error: {event}")
        return False


class SpeechmaticsEngine(WebSocketEngine):
    """Speechmatics Real-Time v2 websocket (design §4)."""

    name = "speechmatics"
    target_rate = 16000
    # EndOfTranscript is a hard terminal: nothing follows it, no drain needed.

    def __init__(self, api_key: str, connect=None):
        super().__init__(api_key, connect=connect)
        self._seq_no = 0  # binary AddAudio frames sent, for EndOfStream

    def _url(self) -> str:
        return SPEECHMATICS_REALTIME_URL

    def _headers(self) -> dict:
        return {"Authorization": f"Bearer {self._api_key}"}

    def _open_messages(self) -> list:
        return [json.dumps({
            "message": "StartRecognition",
            "audio_format": {
                "type": "raw",
                "encoding": "pcm_s16le",
                "sample_rate": self.target_rate,
            },
            "transcription_config": {"language": "en", "enable_partials": True},
        })]

    def _encode_chunk(self, pcm: bytes) -> bytes:
        self._seq_no += 1
        return pcm  # binary AddAudio frame

    def _end_messages(self) -> list:
        return [json.dumps({"message": "EndOfStream", "last_seq_no": self._seq_no})]

    def _handle_event(self, raw) -> bool:
        if isinstance(raw, (bytes, bytearray)):
            return False
        event = json.loads(raw)
        mtype = event.get("message", "")
        if mtype == "AddPartialTranscript":
            self._emit_partial(event.get("metadata", {}).get("transcript", ""))
        elif mtype == "AddTranscript":
            self._emit_final(event.get("metadata", {}).get("transcript", ""))
        elif mtype == "EndOfTranscript":
            return True
        elif mtype == "Error":
            raise RuntimeError(
                f"speechmatics error: {event.get('type')}: {event.get('reason')}"
            )
        return False
